File: to_do_script.js
// Get DOM elements
const todoInput = document.querySelector(".todo-input");
const todoButton = document.querySelector(".todo-button");
const activeTasksContainer = document.querySelector(".active-tasks");
const completedTasksContainer = document.querySelector(".completed-tasks");

let tasks = [];
let taskId = 1;

// Event listeners
todoButton.addEventListener("click", addTask);
todoInput.addEventListener("keydown", function (event) {
  if (event.key === "Enter") {
    event.preventDefault();
    addTask();
  }
});

activeTasksContainer.addEventListener("click", handleTaskAction);
completedTasksContainer.addEventListener("click", handleTaskAction);

// Add a new task
function addTask() {
  const taskName = todoInput.value.trim();

  if (taskName === "") {
    return;
  }

  const newTask = {
    id: taskId,
    name: capitalizeFirstLetter(taskName),
    completed: false,
  };

  tasks.push(newTask);
  renderTasks();

  todoInput.value = "";
  taskId++;
}

// Handle task actions (complete/undo/delete)
function handleTaskAction(event) {
  const target = event.target;
  const taskId = parseInt(target.dataset.taskId);

  if (target.classList.contains("complete-button")) {
    completeTask(taskId);
  } else if (target.classList.contains("undo-button")) {
    undoTask(taskId);
  } else if (target.classList.contains("delete-button")) {
    deleteTask(taskId);
  }
}

// Complete a task
function completeTask(taskId) {
  const task = tasks.find((task) => task.id === taskId);

  if (task) {
    task.completed = true;
    renderTasks();
  }
}

// Undo a completed task
function undoTask(taskId) {
  const task = tasks.find((task) => task.id === taskId);

  if (task) {
    task.completed = false;
    renderTasks();
  }
}

// Delete a task
function deleteTask(taskId) {
  tasks = tasks.filter((task) => task.id !== taskId);
  renderTasks();
}

// Render tasks
function renderTasks() {
  activeTasksContainer.innerHTML = "";
  completedTasksContainer.innerHTML = "";

  tasks.forEach((task) => {
    const taskRow = document.createElement("tr");

    const taskIdCell = document.createElement("td");
    taskIdCell.innerText = task.id;
    taskRow.appendChild(taskIdCell);

    const taskNameCell = document.createElement("td");
    taskNameCell.innerText = task.name;
    taskRow.appendChild(taskNameCell);

    const actionsCell = document.createElement("td");
    if (task.completed) {
      const undoButton = createUndoButton(task.id);
      actionsCell.appendChild(undoButton);
    } else {
      const completeButton = createCompleteButton(task.id);
      const deleteButton = createDeleteButton(task.id);
      actionsCell.appendChild(completeButton);
      actionsCell.appendChild(deleteButton);
    }
    taskRow.appendChild(actionsCell);

    if (task.completed) {
      completedTasksContainer.appendChild(taskRow);
    } else {
      activeTasksContainer.appendChild(taskRow);
    }
  });
}

// Create complete button
function createCompleteButton(taskId) {
  const completeButton = document.createElement("button");
  completeButton.innerText = "Complete";
  completeButton.classList.add("complete-button");
  completeButton.setAttribute("data-task-id", taskId);
  return completeButton;
}

// Create undo button
function createUndoButton(taskId) {
  const undoButton = document.createElement("button");
  undoButton.innerText = "Undo";
  undoButton.classList.add("undo-button");
  undoButton.setAttribute("data-task-id", taskId);
  return undoButton;
}

// Create delete button
function createDeleteButton(taskId) {
  const deleteButton = document.createElement("button");
  deleteButton.innerText = "Delete";
  deleteButton.classList.add("delete-button");
  deleteButton.setAttribute("data-task-id", taskId);
  return deleteButton;
}

// Capitalize first letter of a string
function capitalizeFirstLetter(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}

File: to_do_script.py
from tkinter import *

tasks = []
task_id = 1

def add_task():
    task_name = todo_input.get().strip()

    if task_name == "":
        return

    new_task = {
        "id": task_id,
        "name": task_name.capitalize(),
        "completed": False
    }

    tasks.append(new_task)
    render_tasks()

    todo_input.delete(0, END)
    task_id += 1

def complete_task(task_id):
    task = next((task for task in tasks if task["id"] == task_id), None)

    if task:
        task["completed"] = True
        render_tasks()

def undo_task(task_id):
    task = next((task for task in tasks if task["id"] == task_id), None)

    if task:
        task["completed"] = False
        render_tasks()

def delete_task(task_id):
    global tasks
    tasks = [task for task in tasks if task["id"] != task_id]
    render_tasks()

def render_tasks():
    active_tasks_container.delete(*active_tasks_container.get_children())
    completed_tasks_container.delete(*completed_tasks_container.get_children())

    for task in tasks:
        task_row = (task["id"], task["name"])

        if task["completed"]:
            completed_tasks_container.insert("", END, values=task_row + ("",))
        else:
            active_tasks_container.insert("", END, values=task_row + (create_complete_button(task["id"]), create_delete_button(task["id"])))

def create_complete_button(task_id):
    return Button(text="Complete", command=lambda: complete_task(task_id))

def create_delete_button(task_id):
    return Button(text="Delete", command=lambda: delete_task(task_id))

# Create the main window
root = Tk()
root.title("Todo App")

# Create the input field
todo_input = Entry(root, width=30)
todo_input.grid(row=0, column=0, padx=10, pady=10)

# Create the add button
add_button = Button(root, text="Add Task", command=add_task)
add_button.grid(row=0, column=1, padx=10, pady=10)

# Create the active tasks container
active_tasks_container = Treeview(root, columns=("ID", "Task", "Actions"))
active_tasks_container.heading("#0", text="", anchor=W)
active_tasks_container.column("#0", width=0, stretch=NO)
active_tasks_container.heading("ID", text="ID", anchor=W)
active_tasks_container.heading("Task", text="Task", anchor=W)
active_tasks_container.heading("Actions", text="Actions", anchor=W)
active_tasks_container.column("Actions", width=100, anchor=W)
active_tasks_container.grid(row=1, column=0, padx=10, pady=10)

# Create the completed tasks container
completed_tasks_container = Treeview(root, columns=("ID", "Task", "Actions"))
completed_tasks_container.heading("#0", text="", anchor=W)
completed_tasks_container.column("#0", width=0, stretch=NO)
completed_tasks_container.heading("ID", text="ID", anchor=W)
completed_tasks_container.heading("Task", text="Task", anchor=W)
completed_tasks_container.heading("Actions", text="Actions", anchor=W)
completed_tasks_container.column("Actions", width=100, anchor=W)
completed_tasks_container.grid(row=1, column=1, padx=10, pady=10)

root.mainloop()
